File: shared/src/agent.ts
import type { DataTick } from "./data.js";
import type { Anomaly, AnomalyFeedback } from "./anomaly.js";

export type SessionKind = "monitor" | "subagent" | "improvement";

export type SessionMeta = {
  id: string;
  startedAt: number;
  endedAt?: number;
  kind: SessionKind;
  parentSessionId?: string;
  tokenCount: number;
};

export type AgentMessageRole = "user" | "assistant" | "system";

export type AgentMessage = {
  role: AgentMessageRole;
  content: string;
  model?: string;
  usage?: { input: number; output: number };
  timestamp: number;
};

export type CycleState = {
  cycleId: string;
  sessionId: string;
  batchNumber: number;
  tickCount: number;
  anomaliesDetected: number;
  startedAt: number;
};

export type AgentState = "idle" | "running" | "paused" | "error";

export type AgentStatus = {
  state: AgentState;
  currentSessionId?: string;
  currentCycleId?: string;
  totalCycles: number;
  totalAnomalies: number;
  uptime: number;
  lastError?: string;
};

export type AgentActivityType =
  | "cycle_start"
  | "cycle_end"
  | "anomaly_detected"
  | "memory_flush"
  | "compaction"
  | "subagent_spawn"
  | "feedback_processed"
  | "rule_evolved"
  | "error";

export type AgentActivity = {
  type: AgentActivityType;
  message: string;
  timestamp: number;
  data?: Record<string, unknown>;
};

export type SessionTranscriptEntry =
  | { type: "session"; version: number; id: string; timestamp: string; kind: SessionKind }
  | { type: "data_tick"; source: string; payload: DataTick }
  | { type: "message"; message: AgentMessage }
  | { type: "anomaly"; anomaly: Anomaly }
  | { type: "feedback"; feedback: AnomalyFeedback };

File: src-tauri/src/types/agent.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AgentState {
    Idle,
    Running,
    Paused,
    Error,
    Unhealthy,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentStatus {
    pub state: AgentState,
    pub current_session_id: Option<String>,
    pub current_cycle_id: Option<String>,
    pub total_cycles: u64,
    pub total_anomalies: u64,
    pub uptime: u64,
    pub last_error: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AgentActivityType {
    CycleStart,
    CycleEnd,
    AnomalyDetected,
    MemoryFlush,
    Compaction,
    SubagentSpawn,
    FeedbackProcessed,
    RuleEvolved,
    Error,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentActivity {
    #[serde(rename = "type")]
    pub activity_type: AgentActivityType,
    pub message: String,
    pub timestamp: u64,
    pub data: Option<std::collections::HashMap<String, serde_json::Value>>,
}
